### PR TITLE
refactor(facade): remove simple facade/lang functions

### PR DESCRIPTION
**Pull request by jasonchoimtt**
- [x] Inline uses in `@angular` packages
- [x] Simplify `isPresent/isBlank predicate` for objects
- [x] Inline uses in `playground`
- [x] Remove imports
- [x] Remove implementation in facade

**Please check if the PR fulfills these requirements**
- The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit-message-format
- Tests for the changes have been added (for bug fixes / features)
- Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x")

```
Refactoring (no functional changes, no api changes)
```

**Does this PR introduce a breaking change?** (check one with "x")

```
No
```

### CODE REVIEW

**Comment by tolemac** [comment]
I'm courious, why do you remove `isPresent` function? 
I like to use it, are you removing it from entire source code?  Why?

**Comment by jasonchoimtt** [comment]
The facade functions exist mainly because of the need to support both TypeScript and Dart. Now the Dart version is maintained separately, so it is no longer needed.

**Comment by vicb** [comment]
> I'm courious, why do you remove isPresent function? 

kind of agree here, let's discuss this tomorrow

**Comment by Fank** [comment]
@vicb any progress here?

**Comment by vicb** [comment]
not that I know

**Comment by jasonchoimtt** [comment]
The main motivation was just to remove as much facade as possible. However, other than that, I do think removing them is a good idea. iirc there were around 1/3 uses of `isPresent/isBlank` being on object types, which could be simplified to just truthy checks.

I can redo this PR if this is the way to go. There is another thing that I want to change though: to keep `isPromise` in `@angular/core`, so that we can make use of type guards, i.e. using

```
export function isPromise(x: any): x is Promise { ... }
```

will avoid `as any as Promise` casts.

**Comment by vicb** [comment]
> I can redo this PR if this is the way to go.

I think it is, please ping @IgorMinar or @mhevery 

> `export function isPromise(x: any): x is Promise { ... }`

should probably be `export function isPromise(x: any): x is PromiseLike { ... }`
I would be fine with keeping the `isPromise` name - the other option is to use `isPromiseLike`

**Comment by IgorMinar** [comment]
yes! we should still do this, but we should review the inlining of `is*` methods. I like the idea of ensuring that the inlining preserves or improves type guards

inlining stuff using instanceof for objects and typeof for primitives automatically activates ad type guard

based on my quick review, only isDate and isPromise might need to be specialized. isDate contains a NaN check, which looks bogus - it was introduced by someone adding DatePipe. We should move the isNaN check into the `supports` method of the pipe.

we should keep isPromise and move it to the core and reexport it via core_private for other packages.

**Comment by IgorMinar** [comment]
the pr still needs `gulp format` and instead of `chore` it should be marked as `refactor` in the commit message. otherwise lgtm

**Comment by jasonchoimtt** [comment]
@IgorMinar I'm actually not done, but if you want to merge the `isPromise` part first I'm fine too.

@vicb I couldn't make `PromiseLike` work since it is not assignable to `Promise` (the former not having `.catch` afaik). Alternative might be to use `PromiseLike` in some parts of the codebase too.

File: modules/@angular/common/src/pipes/async_pipe.ts [hunk 1]
```diff
@@ -8,7 +8,8 @@
 
 import {ChangeDetectorRef, OnDestroy, Pipe, WrappedValue} from '@angular/core';
 import {EventEmitter, Observable} from '../facade/async';
-import {isBlank, isPresent, isPromise} from '../facade/lang';
+import {isBlank, isPresent} from '../facade/lang';
```
Comment (vicb) [inline modules/@angular/common/src/pipes/async_pipe.ts]: could you move isBlank/isPresent to util also ?

File: modules/@angular/core/src/util/lang.ts [hunk 2]
```diff
@@ -0,0 +1,13 @@
+/**
+ * @license
+ * Copyright Google Inc. All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+export function isPromise(obj: any): obj is Promise<any> {
```
Comment (vicb) [inline modules/@angular/core/src/util/lang.ts]: probably good enough for now but we also probably want a `isPromiseLike` (=thenable) at some point.

When should then `isPromise` -> `isPromise` wherever possible (ie `catch` no called)

File: modules/@angular/core/test/util/lang_spec.ts [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+ */
+import {isPromise} from '@angular/core/src/util/lang';
+
+export function main() {
+  describe('isPromise', () => {
+    it('should be true for native Promises',
+       () => expect(isPromise(Promise.resolve(true))).toEqual(true));
+
+    it('should be true for thenables',
+       () => expect(isPromise({then: function() {}})).toEqual(true));
```
Comment (vicb) [inline modules/@angular/core/test/util/lang_spec.ts]: `() => {}`

**Comment by jasonchoimtt** [comment]
are we still inlining `isPresent/isBlank`?

**Comment by vicb** [comment]
They're useful -1 for inlining everywhere

/cc @pkozlowski-opensource

**Comment by jasonchoimtt** [comment]
Point taken. I think we should just merge the `isPromise` part first then, since the scale of `isPresent/isBlank` is much bigger.

**Comment by alexeagle** [comment]
@jasonchoimtt I may pick up some of the facade refactoring also - do you have a link to the TS program you used to analyze the usages and produce the replacements?

**Comment by jasonchoimtt** [comment]
Yes, here it is:

https://gist.github.com/jasonchoimtt/6b8a49455f337f35d3b7c461deff5e7a#file-facade-killer-ts

It works for functions with a single return statement, but may do weird things with respect to operator precedence and multiple statements.

The code is rather messy, but it should work if you keep the cases simple. Also make sure you run `clang-format` afterwards.

**Comment by alexeagle** [comment]
thanks!

**Pull request merged**